src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -0,0 +1,229 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Internal;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding
+{
+    /// <summary>
+    /// A factory for <see cref="IModelBinder"/> instances.
+    /// </summary>
+    public class ModelBinderFactory : IModelBinderFactory
+    {
+        private readonly IModelMetadataProvider _metadataProvider;
+        private readonly IModelBinderProvider[] _providers;
+
+        private readonly ConcurrentDictionary<object, IModelBinder> _cache;
+
+        /// <summary>
+        /// Creates a new <see cref="ModelBinderFactory"/>.
+        /// </summary>
+        /// <param name="metadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
+        /// <param name="options">The <see cref="IOptions{TOptions}"/> for <see cref="MvcOptions"/>.</param>
+        public ModelBinderFactory(IModelMetadataProvider metadataProvider, IOptions<MvcOptions> options)
+        {
+            _metadataProvider = metadataProvider;
+            _providers = options.Value.ModelBinderProviders.ToArray();
+
+            _cache = new ConcurrentDictionary<object, IModelBinder>(ReferenceEqualityComparer.Instance);
+        }
+
+        /// <inheritdoc />
+        public IModelBinder CreateBinder(ModelBinderFactoryContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            IModelBinder binder;
+            if (context.CacheToken != null && _cache.TryGetValue(context.CacheToken, out binder))
+            {
+                return binder;
+            }
+
+            var providerContext = new DefaultModelBinderProviderContext(this, context);
+            binder = CreateBinderCore(providerContext, context.CacheToken);
+            if (binder == null)
+            {
+                var message = Resources.FormatCouldNotCreateIModelBinder(providerContext.Metadata.ModelType);
+                throw new InvalidOperationException(message);
+            }
+
+            if (context.CacheToken != null)
+            {
+                _cache.TryAdd(context.CacheToken, binder);
+            }
```

Comment:
> Shouldn't we cache the computed binder for TryUpdateModel[Async]() scenarios?

It would be nice to do that - sure. Maybe using the metadata as the token?